:sparkles: fallback component

ptitFicus opened the pull request:
fixes #49 

Hi,

I adapted code of #66 (initially made for #49) to the new version of this HOC.

I also:

- Changed `isInState` name to `hasStatus`: it may still be unclear but a last it's not confusing with react `state` anymore
- Factorized the whole coloration code in a util folder. It avoids duplication but util folder is not necessarily a very good pattern: any idea welcome :blush:

The is some linting issues, since I call `setState` in `componentDidMount` directly :scream:, however the old `componentDidMount` dit already call `setState`, it was just hidden in the `setColor` function. We should either change linting rules or find a new color strategy.

README is missing an explicative section about this improvement, but I would like to make sure we all agree on this before writting it.

Let's talk about this again! :smiley:

fabienjuif commented:
Take care, you don't use you SSH key to push those commits.
So your name is the not the right one on commits ;)

fabienjuif commented on src/ErrorCross/ErrorCross.jsx:
```diff
@@ -0,0 +1,64 @@
+    return (
+      <div
+        title={message}
+        ref={(c) => { this.div = c }}
```
Now I prefer to do this : 
```es6
class extends Component {
  attach: () => {
    const newColor = findCorrectColor(this.div)
     if (this.state.color !== newColor) {
       this.setState({ color: newColor })
     }
  }

  render () {
    return <div ref={this.attach} />
  }
}
```

So I am sure of the order, and I don't rely on the `this` context

fabienjuif commented on src/ErrorCross/ErrorCross.jsx:
```diff
@@ -0,0 +1,64 @@
+        >
+          <path
+            stroke={color}
+            strokeWidth="3.03754568"
```
What a precision :)

fabienjuif commented on src/ErrorCross/ErrorCross.jsx:
```diff
@@ -0,0 +1,64 @@
+  }
+}
+
+const { string, object } = PropTypes
```
clever one :p

frinyvonnick replied:
👏

fabienjuif commented on src/TailSpin/TailSpin.jsx:
```diff
@@ -1,42 +1,21 @@
-      this.setState({
-        color: color.toHexString(),
-      })
+    const newColor = findCorrectColor(this.svg)
```
Same comment as before (`attach` method)

fabienjuif commented on src/core.jsx:
```diff
@@ -101,7 +96,9 @@ export default (
 
       render() {
-        if (!isLoaded(this.props, this.state, this.context)) {
+        if (isInError(this.props, this.state, this.context)) {
```
We will avoid negative test and `else` :  
```es6
if (isInError) return <ErrorIndicator ...
if (isLoaded) return <ComposedComponent ...
return <LoadingIndicator ...
```

fabienjuif commented on src/core.jsx:
```diff
@@ -22,41 +22,36 @@ const getDisplayName = c => c.displayName || c.name || 'Component'
-        .map(p => Boolean(props[p]))
-        .reduce((allProps, currentProp) => allProps && currentProp)
+    if (Array.isArray(prop)) {
+      const boolProps = prop.map(p => Boolean(props[p]))
```
We should avoir this to kind of writing (I know this was the case before, you can open an issue if you want to argue).
But here in `hasStatus` we have `!!` technique and `Boolean` technique

fabienjuif commented on src/index.spec.js:
```diff
@@ -125,4 +126,75 @@ describe('hoc-react-loader', () => {
     })
   })
+
+  describe('error parameter', () => {
```
Nice 👍 

What do you think of this new style of writing our test here ?

fabienjuif commented on src/utils/colorUtil.js:
```diff
@@ -0,0 +1,27 @@
+  return window.getComputedStyle(node, null).getPropertyValue('background-color')
+}
+
+export const initialColor = '#cecece'
```
Do we have to use `UPPER_CASE` convention for constants ?

fabienjuif commented:
I renamed your PR so the squash/merge would have a better message to me

fabienjuif commented:
And I update your first post to add the `fixes` triggered word

fabienjuif commented:
Nice job @ptitFicus 👍 
Do you want to rebase your branch so you can set a good author ? 

 - `git rebase -i master`
 - For each commit:
    * `git commit --amend --author 'Benjamin CAVY <benjamin.cavy@gmail.com>'`
    * `git rebase --continue`

ptitFicus commented:
Thanks for the advice @fabienjuif ! I was about to give up on this author mess 😄 

I added a dedicated section to the README + I updated project version.

fabienjuif commented on README.md:
```diff
@@ -74,3 +74,23 @@ The default `print` parameter value is `true`. It means that in this example the
 The `print` parameter can also be a function. Then the `props` and `context` are given to it (in this order), and it should return a boolean indicating wether or not the actual component should be displayed.
+
+### Error handling
+The `error` parameter allow to specify a prop that indicates wether or not a placeholder error component should be displayed in replacement of the real component.
```
`allows`

fabienjuif commented on README.md:
```diff
@@ -74,3 +74,23 @@ The default `print` parameter value is `true`. It means that in this example the
+
+### Error handling
+The `error` parameter allow to specify a prop that indicates wether or not a placeholder error component should be displayed in replacement of the real component.
+It's usefull when data that are mandatory for the correct display of a component are missing.
```
I think this is @hgwood, but someone once tells me to use `required` instead of `mandatory`.
I can't remember why tho.

hgwood replied:
It's just way more common to use "required". "mandatory" is more formal. In general shorter words are preferred (eg speed vs velocity, give up vs abandon, build vs construct...). I often use the shortcut that if it's French-sounding, then it's formal :).

fabienjuif commented on README.md:
```diff
@@ -74,3 +74,23 @@ The default `print` parameter value is `true`. It means that in this example the
+The `error` parameter allow to specify a prop that indicates wether or not a placeholder error component should be displayed in replacement of the real component.
+It's usefull when data that are mandatory for the correct display of a component are missing.
+
+Like for the `print` prop, `error` can be a `boolean`, a `string` (referencing a prop name), an array of `string` (an array of prop names) or a `function` (whose result will be converted to `boolean`).
```
`As the ``print`` prop` ?

ptitFicus replied:
Are you sure about this one ? English is not my strong suit however I think using "as" would mean that `error` could be use in place of `print`.

fabienjuif replied:
ok

fabienjuif commented on examples/src/components/Examples/Examples.jsx:
```diff
@@ -106,7 +106,7 @@ const Examples = ({ style, className }) => (
       className={styles.example}
       link="ErrorIndicator"
-      code="export&nbsp;default loader(Base)"
+      code="export&nbsp;default loader()(Base)"
```
👍
